### PR TITLE
Update Xively makefile to use wolfSSL release version 3.10.2-stable

### PR DESCRIPTION
**Pull request by DellaBitta**
[ Description ]
Updated the build environments to checkout a newer version of WolfSSL's TLS library.

[ JIRA ]
XCL-2946
Update Build Script to use WolfSSL v3.10.2

[ Reviewer ]

[ QA ]
Built and executed demos on macOSX with POSIX.
Tested build process and executed the demos for the CC3200 tutorial and the CC3200 tutorial with enhanced security
Tested build process and executed demo on the STM32F4 Nuceo Wired example tutorial, which requires WolfSSL to be built.

[ Release Notes ]
Updated support to build WolfSSL's TLS library veriosn 3.10.2-stable.

### CODE REVIEW

File: examples/cc3200/xively_demo/.project [hunk 1]
```diff
@@ -59,7 +59,7 @@
 	<variableList>
 		<variable>
 			<name>CC3200_SDK_ROOT</name>
-			<value>file:/C:/ti/tirex-content/CC3200SDK_1.1.0/cc3200-sdk</value>
+			<value>file:/C:/ti/CC3200SDK_1.1.0/cc3200-sdk</value>
```
Comment (DellaBitta) [inline examples/cc3200/xively_demo/.project]: new default path for these installations have removed hte tirex-content directory.

File: examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject [hunk 2]
```diff
@@ -187,6 +187,45 @@
...
+									<listOptionValue builtIn="false" value="NO_SERVER"/>
+									<listOptionValue builtIn="false" value="USER_TIME"/>
+									<listOptionValue builtIn="false" value="HAVE_ECC"/>
+									<listOptionValue builtIn="false" value="HAVE_TLS_EXTENSIONS"/>
+									<listOptionValue builtIn="false" value="HAVE_AESGCM"/>
+									<listOptionValue builtIn="false" value="ALT_ECC_SIZE"/>
+									<listOptionValue builtIn="false" value="NO_FILESYSTEM"/>
+									<listOptionValue builtIn="false" value="WC_RSA_BLINDING"/>
+									<listOptionValue builtIn="false" value="WOLFSSL_NO_SOCK"/>
 								</option>
```
Comment (DellaBitta) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject]: I had to add the wolfSSL preprocessor definitions to the main project for the sources to compile correctly, as the headers require new special toggles.

File: make/mt-config/mt-tls-wolfssl.mk [hunk 3]
```diff
@@ -17,3 +17,9 @@ XI_CONFIG_FLAGS += -DXI_TLS_OCSP_STAPLING
 # XI_CONFIG_FLAGS += -DXI_TLS_OCSP
 
 XI_CONFIG_FLAGS += -DXI_TLS_LIB_WOLFSSL
+
+XI_CONFIG_FLAGS += -DWC_RSA_BLINDING
+
+XI_CONFIG_FLAGS += -DNO_FILESYSTEM
+
+XI_CONFIG_FLAGS += -DWOLFSSL_NO_SOCK
```
Comment (DellaBitta) [inline make/mt-config/mt-tls-wolfssl.mk]: New WolfSSL preprocessor definitions required for compilation.   This is for  our makefile environment.

File: examples/cc3200/xively_demo/.project (hunk 1 above)
Comment (atigyi) [inline examples/cc3200/xively_demo/.project]: 👍

File: examples/stm32/WolfSSL/.project [hunk 4]
```diff
@@ -320,5 +320,10 @@
 			<type>1</type>
 			<locationURI>PARENT-3-PROJECT_LOC/src/import/tls/wolfssl/ctaocrypt/src/wolfcrypt_last.c</locationURI>
 		</link>
+		<link>
+			<name>wolfmath.c</name>
+			<type>1</type>
+			<locationURI>PARENT-3-PROJECT_LOC/src/import/tls/wolfssl/wolfcrypt/src/wolfmath.c</locationURI>
```
Comment (atigyi) [inline examples/stm32/WolfSSL/.project]: as I see this is a new file in the new wolfSSL version, right?

File: examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject (hunk 2 above)
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject]: hm, at first sight it is weird why main project requires wolfssl defines. The main project shouldn't even know about security. It shouldn't know whether the connection is secure or not.

File: src/import/tls/download_and_compile_wolfssl.sh [hunk 5]
```diff
@@ -54,7 +54,7 @@ then
 fi
 git clone https://github.com/wolfSSL/wolfssl
 cd wolfssl
-git checkout tags/v3.9.6
-(autoreconf --install && ./configure `cat ../wolfssl.conf` && make)
+git checkout tags/v3.10.2-stable
```
Comment (atigyi) [inline src/import/tls/download_and_compile_wolfssl.sh]: 👍

File: examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject (hunk 2 above)
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject]: which headers required wolfSSL toggles? I was able to build STM32 eth project with removing these wolfSSL defines. Although I haven't tested it yet.

File: examples/stm32/WolfSSL/.settings/language.settings.xml [hunk 6]
```diff
@@ -5,7 +5,7 @@
 			<provider copy-of="extension" id="org.eclipse.cdt.ui.UserLanguageSettingsProvider"/>
 			<provider-reference id="org.eclipse.cdt.core.ReferencedProjectsLanguageSettingsProvider" ref="shared-provider"/>
 			<provider-reference id="org.eclipse.cdt.managedbuilder.core.MBSLanguageSettingsProvider" ref="shared-provider"/>
-			<provider class="fr.ac6.mcu.ide.build.CrossBuiltinSpecsDetector" console="false" env-hash="757024362449438222" id="fr.ac6.mcu.ide.build.CrossBuiltinSpecsDetector" keep-relative-paths="false" name="Ac6 SW4 STM32 MCU Built-in Compiler Settings" parameter="${COMMAND} ${FLAGS} -E -P -v -dD &quot;${INPUTS}&quot;" prefer-non-shared="true">
+			<provider class="fr.ac6.mcu.ide.build.CrossBuiltinSpecsDetector" console="false" env-hash="-759066157078245430" id="fr.ac6.mcu.ide.build.CrossBuiltinSpecsDetector" keep-relative-paths="false" name="Ac6 SW4 STM32 MCU Built-in Compiler Settings" parameter="${COMMAND} ${FLAGS} -E -P -v -dD &quot;${INPUTS}&quot;" prefer-non-shared="true">
```
Comment (atigyi) [inline examples/stm32/WolfSSL/.settings/language.settings.xml]: 🤜 🤕 🤛  fight!!

File: examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject [hunk 7]
```diff
@@ -4,9 +4,9 @@
 		<cconfiguration id="fr.ac6.managedbuild.config.gnu.cross.exe.debug.667843212">
 			<storageModule buildSystemId="org.eclipse.cdt.managedbuilder.core.configurationDataProvider" id="fr.ac6.managedbuild.config.gnu.cross.exe.debug.667843212" moduleId="org.eclipse.cdt.core.settings" name="IKS01A2_Debug">
 				<macros>
-					<stringMacro name="XI_STMCUBE_PATH" type="VALUE_PATH_DIR" value="/Users/palantir/Work/Xively/stm32f4xx/STM32Cube_FW_F4_V1.16.0"/>
+					<stringMacro name="XI_STMCUBE_PATH" type="VALUE_PATH_DIR" value="/Users/ddellabitta/dev/st/STM32Cube_FW_F4_V1.16.0"/>
 					<stringMacro name="XI_CLIENT_PATH" type="VALUE_PATH_DIR" value="../../../../../.."/>
-					<stringMacro name="XI_MEMS_STMCUBE_PATH" type="VALUE_PATH_DIR" value="/Users/palantir/Work/Xively/stm32f4xx/STM32CubeExpansion_MEMS1_V4.0.0"/>
+					<stringMacro name="XI_MEMS_STMCUBE_PATH" type="VALUE_PATH_DIR" value="/Users/ddellabitta/dev/st/STM32CubeExpansion_MEMS1_V4.0.0"/>
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject]: 🤜 🤕 🤛  fight!!

File: make/mt-config/mt-tls-wolfssl.mk (hunk 3 above)
Comment (atigyi) [inline make/mt-config/mt-tls-wolfssl.mk]: I've removed these defines and the posix version still compiles and links. Are these really necessary?

Comment (atigyi) [inline make/mt-config/mt-tls-wolfssl.mk]: Also removed from the STM32's eth WolfSSL project, still compiles, links. I didn't test it on device.

File: make/mt-os/mt-cc3200.mk [hunk 8]
```diff
@@ -5,41 +5,59 @@
 
 include make/mt-os/mt-os-common.mk
 
-###
-## COMPILER NAME
-###
-COMPILER ?= arm_15.12.3.LTS
+ifdef XI_TRAVIS_BUILD
+	# For Xively C Client Continuous Integration only.
+	COMPILER ?= arm_15.12.3.LTS
```
Comment (atigyi) [inline make/mt-os/mt-cc3200.mk]: shouldn't we upgrade Travis' compiler instead?

Comment (atigyi) [inline make/mt-os/mt-cc3200.mk]: I guess it is "just" adding the new path (similar to xively-client-artifactory/ti/ccsv6/tools/compiler/arm_15.12.3.LTS) to the artifactory. Although it has the be the linux version of the ccsv7.

Comment (atigyi) [inline make/mt-os/mt-cc3200.mk]: maybe this is a separate task / story: to update travis' CC3200 compiler, so I guess we can skip this here

File: make/mt-os/mt-cc3200.mk [hunk 9]
```diff
@@ -5,41 +5,59 @@
...
 ###
+
+#
+## MAC HOST OS
+#
 ifeq ($(XI_HOST_PLATFORM),Darwin)
 	# osx cross-compilation downloads
 
-	XI_CC3200_PATH_CCS_TOOLS ?= /Applications/ti/ccsv6/tools
+	XI_CC3200_PATH_CCS_TOOLS ?= /Applications/ti/ccsv7/tools
```
Comment (atigyi) [inline make/mt-os/mt-cc3200.mk]: 👍

**Comment by atigyi** [comment]
I have 2 concerns:
- I am not convinced about the necessity of the wolfSSL define introduction neither in STM32's main project nor in our mt-tls-wolfssl.mk, maybe I am wrong but I was able to build all without the new defines for both posix and stm32
- I am not sure, just asking: doesn't cc3220sf wolfssl project needs update under the example/cc3220sf/wolfssl_ccs directory?

File: examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject (hunk 2 above)
Comment (DellaBitta) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_eth/SW4STM32/STM32F429ZI_NUCLEO_144/.cproject]: I'm not sure, but as my previous note stated, the ST Project, Build and Tutorial process are broken.  I've created a story to sort it all out as it's not part of this PR:  https://jira.3amlabs.net/browse/XCL-3037

File: make/mt-config/mt-tls-wolfssl.mk (hunk 3 above)
Comment (DellaBitta) [inline make/mt-config/mt-tls-wolfssl.mk]: did you remove the wolfssl directory and checkout the new wolfssl version?

File: examples/stm32/WolfSSL/.project (hunk 4 above)
Comment (DellaBitta) [inline examples/stm32/WolfSSL/.project]: not sure.  Again, don't know if I'm fighting broken project or new wolfssl version here, but it failed to compile without this.

File: make/mt-os/mt-cc3200.mk (hunk 8 above)
Comment (DellaBitta) [inline make/mt-os/mt-cc3200.mk]: Yes. I tried to do it quickly but I didn't have a linux machine to run the binary installation process, so this was my short term step.

**Comment by DellaBitta** [comment]
I would double check that you actually have the 3.10.2 version that you've built, because I had problems with every platform.

File: examples/stm32/WolfSSL/.project (hunk 4 above)
Comment (DellaBitta) [inline examples/stm32/WolfSSL/.project]: given the fact that the project files work now
I'm guessing it's a new file. :)

Comment (atigyi) [inline examples/stm32/WolfSSL/.project]: ok, it is a new file came into wolfSSL on 2016 dec 21, right after v3.10.0-stable

File: make/mt-os/mt-cc3200.mk (hunk 8 above)
Comment (atigyi) [inline make/mt-os/mt-cc3200.mk]: 👍

**Comment by atigyi** [comment]
💯 Approved, great job!!
